### PR TITLE
a simple implementation of highlighting

### PR DESCRIPTION
**Pull request by cnwangjie**
@wlach Hi, I tried a very simple implementation of highlighting feature. What do you think?

I have considered setting options for which type to choose the highlighted points. If we take data as the option, because the data we saved is a copy of the data object instead of a reference to the data, we need to implement a method to determine whether the data corresponds to. If we use the index as an option, then the index of the data will change when a part of the data is filtered. So I think it's better to use a callback function, which allows the user easily to use another library such as `isEqual` or `isMatch` methods of Lodash to select the data that needs to be highlighted.

Following is the effect of setting `highlight: (d, i) => d.y > 120 && d.y < 180`

image

### CODE REVIEW

**Comment by wlach** [comment]
I quite like this approach. @hamilton, any thoughts? If he's good with it, we should add an example or two.

**Comment by hamilton** [comment]
hmmm ... wouldn't this be better if we used `color` as the option here instead of highlight, and let `color` (in this case) be a function that operates on the data point itself, as per your example? That' lend itself to greater flexibility generally. I'm still thinking about the overall primitives here, though. Circles are different than lines. You can pass in an array of arrays, and perhaps the idea would be that each array gets its own color type. `color` on lines operates on the whole array of data, not a particular point, but I can see in this case wanting to do both, maybe. Perhaps the arguments for `color`, as a function, would be `d, i, j` where `d` is the data point, `i` is the elem within the array, and `j` is which set of array (so if I pass in an array containing two arrays of data to be rendered to the points, `j` would refer to which of the top-level arrays I'm referring to).

I think I'd want to say

```javascript
MG.data_graphic({
   ...
   color: (d,i)=> (d.y > 120 && d.y < 180) ? 'yellow' : undefined // maybe undefined defaults to whatever the usual color is?
})
```

Thoughts here?

**Comment by cnwangjie** [comment]
@hamilton But now the point chart can already set the color of the point. This may be repeated. How do you think of this?

reference: https://www.metricsgraphicsjs.org/examples.htm#experimental

If the user wants to set some data points to a color, they just need to add an attribute to the data and set the attribute as `color_accessor`.

**Comment by hamilton** [comment]
Hey @cnwangjie - 

Ah, I misunderstood. `highlight` changes one aspect of the circle size - an additional, bigger border. It does not affect color.

How does this interact with the `size_accessor` option? If I set both `size_accessor` and `highlight`, how is that intended to work?

**Comment by hamilton** [comment]
Otherwise, I think this looks good. :) Great work!

**Comment by cnwangjie** [comment]
I imagined the effect is adding a border to the outer of the point. Following are the effect with `size_accessor` and `highlight` options at the same time. Points of different sizes still displayed different sizes after being highlighted.

image

image

I think this effect is not too bad, but there is a little problem is stroke will make the middle circle deeper (like the following case). But now svg can't control the align of the stroke. (I tried to find out if there is such a property, but [this property is not supported](https://caniuse.com/#search=stroke-align)) 

image

One solution I found was to create a larger circle and set stroke. Following is the effect I achieved by creating more circles. I think it looks a little better than above.

image

What do you think about this? @hamilton @wlach 

And could you give me some advice about organizing new code, if there is a better way?

File: src/js/charts/point.js [hunk 1]
```diff
@@ -195,18 +195,43 @@ function mg_color_point_mouseover({color_accessor, scalefns}, elem, d) {
...
+        }
       } else {
         pts.classed('mg-points-mono', true);
+        if (highlights) {
+          highlights.classed('mg-points-mono', true);
+        }
       }
 
       if (args.size_accessor !== null) {
         pts.attr('r', args.scalefns.sizef);
```
Comment (wlach) [inline src/js/charts/point.js]: Could we do something like:

```js
pts.attr('r', (args.size_accessor !== null) ?  args.scalefns.sizef : args.point_size);
if (highlights) {
  highlights.attr('r', (args.size_accessor !== null) ? (d, i) => args.scalefns.sizef(d, i) + 2 : args.point_size + 2);
}
```

File: src/js/charts/point.js [hunk 2]
```diff
@@ -195,18 +195,43 @@ function mg_color_point_mouseover({color_accessor, scalefns}, elem, d) {
         .attr('cx', args.scalefns.xoutf)
         .attr('cy', d => args.scalefns.youtf(d));
 
+      let highlights
```
Comment (wlach) [inline src/js/charts/point.js]: Missing semicolon here (not sure why the linter didn't pick this up, need to look into that)

File: src/js/charts/point.js [hunk 3]
```diff
@@ -195,18 +195,43 @@ function mg_color_point_mouseover({color_accessor, scalefns}, elem, d) {
...
+        highlights = svg.append('g')
+          .classed('mg-highlight', true)
+          .selectAll('circle')
+          .data(data.filter(args.highlight))
+          .enter().append('circle')
+          .attr('cx', args.scalefns.xoutf)
+          .attr('cy', d => args.scalefns.youtf(d))
+      }
+
       //are we coloring our points, or just using the default color?
```
Comment (wlach) [inline src/js/charts/point.js]: Maybe we could shorten this a bit?

```js
const elements = [pts].concat(highlights ? [highlights, []]);
if (args.color_accessor !== null) {
  elements.forEach(e => e.attr('fill', args.scalefns.colorf) && e.attr('stroke', args.scalefns.colorf);
} else {
  elements.forEach(e => e.classed('mg-points-mono', true));
}
```

File: src/js/charts/point.js [hunk 4]
```diff
@@ -388,7 +374,8 @@ function mg_color_point_mouseover({color_accessor, scalefns}, elem, d) {
     size_domain: null,
     color_domain: null,
     active_point_size_increase: 1,
-    color_type: 'number' // can be either 'number' - the color scale is quantitative - or 'category' - the color scale is qualitative.
+    color_type: 'number', // can be either 'number' - the color scale is quantitative - or 'category' - the color scale is qualitative.
+    highlight: null // callback function to select the point to highlight
```
Comment (wlach) [inline src/js/charts/point.js]: let's change this comment to `// if this callback function returns true, the selected point will be highlighted`

**Comment by wlach** [comment]
I'm going to land this, it looks good to me. Thanks @cnwangjie!

**Pull request merged**